lib/puppet/pops/model/factory.rb

```diff
@@ -89,6 +90,29 @@ def initialize(o, *args)
     end
   end
 
+  def map_offset(model, locator)
+    MAPOFFSET_VISITOR.visit_this_1(self, model, locator)
+  end
+
+  def map_offset_Object(o, locator)
+    o
+  end
+
+  def map_offset_Factory(o, locator)
+    map_offset(o.model, locator)
+  end
+
+  def map_offset_Positioned(o, locator)
+    # Transpose the local offset, length to global "coordinates"
+    global_offset, global_length = locator.to_global(o.offset, o.length)
+
+    # mutate
+    o.instance_variable_set(:'@offset', global_offset)
+    o.instance_variable_set(:'@length', global_length)
+    # Change locator since the positions were transposed to the global coordinates
+    o.instance_variable_set(:'@locator', locator.locator) if locator.is_a? Puppet::Pops::Parser::Locator::SubLocator
```

Comment:
They should really be read-only, and I wish it was possible to set the values when the objects were instantiated. I did consider making the writeable, but opted for the instance_var set instead to prevent accidental changes. The AST is API used by other tools and they should not have the ability to change the locations.